### PR TITLE
Ecosystem explorer

### PR DESCRIPTION
**Pull request by Fox32**
This plugin provides an entry point into the domains that are available in the catalog. The cards allow to drill into domains and further into more detailed levels from there.

It also makes some adjustments to the `ItemCard` from `@backstage/core`.

domains

Right now this is implemented as a new plugin called `ecosystem`. I'm not 100% satisfied with that, but looking for some feedback. There isn't lot of code and it's questionable whether this should be a standalone plugin. But for now I followed the same patterns as the `org` plugin. As this adds a new plugin, I would like to avoid adding it first and later removing it - so we should know what our (long term) plan is here.

We also have other options and I would be open to adjust this PR:

## Integrate into `catalog` plugin

We could just move everything from the `ecosystem` plugin into the `catalog` plugin. It's tightly coupled to the catalog anyway.

## Bring back life into the `explore` plugin

There is already an `explore` plugin that shows a hard-coded static list of Backstage plugins. The plugin looks like this:

image

We could make the explore plugin the entry point into everything that is available inside an organization. For me that would include the domains, tools and a lot more which is organization specific. We can move everything from the `ecosystem` plugin there. I would expect that tools of an organization are surfaced as Backstage plugins, so we could keep the existing list of Backstage plugins from the explore plugin. Not sure how, but that list should be customizable by the integrations of a Backstage instance. But this could follow in a later step. Here is a mock how it could look like:

image

What do you think?

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)

### CODE REVIEW

File: plugins/ecosystem/src/components/DomainExplorerPage/DomainExplorerPage.tsx
Comment (adamdmharvey) [inline plugins/ecosystem/src/components/DomainExplorerPage/DomainExplorerPage.tsx:67]: For an entire Backstage instance to lack domains, I wonder if the lack of domains is less about the mechanics of how to define the Kind, but more a lack of understanding how the model helps map to business benefits?  I might wonder if this link could be more useful being to the model info:
https://backstage.io/docs/features/software-catalog/system-model#ecosystem-modeling

And then maybe what's missing is links from the model page TO the next level of detail, e.g. to the `Kind:domain` and `Kind:system` descriptors?

Comment (Fox32) [reply plugins/ecosystem/src/components/DomainExplorerPage/DomainExplorerPage.tsx:67]: Yeah we could switch that link out

> And then maybe what's missing is links from the model page TO the next level of detail, e.g. to the Kind:domain and Kind:system descriptors?

Yeah, but I don't think that it should be part of this PR. That can be a separate contribution.

Comment (adamdmharvey) [reply plugins/ecosystem/src/components/DomainExplorerPage/DomainExplorerPage.tsx:67]: 100% agree, didn't mean to infer to block this PR. :)

File: plugins/ecosystem/src/plugin.ts
Comment (Fox32) [reply plugins/ecosystem/src/plugin.ts:27]: Thought about it, but @dhenneke and I were not sure if we can do it when the app itself isn't ported yet. Does that work?

**Comment by freben** [comment]
I have to punt on reviewing fully 'til tomorrow, but regarding the package structure:

I spontaneously preferred the idea of extending the `explore` plugin. That was an interesting angle, and one that will only consume one slot in the sidebar - which I think is a pretty important factor in reachability for this feature. There's a big conceptual overlap as well.

@backstage/maintainers what's your take on that?

I think it's worth considering the persona here. Who will this view target?

To me it sounds like a domains view leans toward the discoverer, somebody who is looking for info about a specific larger concept, or perhaps just wants a bird's eye view of what's out there. Especially if we tie documentation and contact info and a lot of stuff like that to the domains. A component owner likely won't visit this one as much. At least if this is the view of "all the domains".

There may be room for a Systems view too, but that one would likely be targeted at you as a person, showing the ones that you (or any of your org ancestor groups upwards) own, and show collated/merged status info for them etc. Where would such a view live? The catalog package? Probably.

**Comment by Fox32** [comment]
The systems view would kind of depend on the use case. If you are drilling down from the top I would see it in the related domain entity page. But as a team member I really want to have a view for all my systems - independent from the domain.

**Comment by freben** [comment]
Mwell one would be the details view of a domain, the other one would be "my systems". Gotta figure out where these fit in

**Comment by freben** [comment]
A conflict has popped up! Will have a look at this now. @benjdlambert could you have a look too so maybe we can get this in for the release tomorrow?

**Comment by freben** [comment]
Oh @Fox32 I realize that we talked about whether this should go into explore ... @backstage/maintainers and yourself, any thoughts about that? I think it would be both a smaller change and a better placement.

**Comment by Fox32** [comment]
Exactly, but I think I'm already at the state that I this should be part of the explore plugin. But I might need some stuff from #4265, not sure if I can finish this till tomorrow.

**Comment by freben** [comment]
OK, don't stress out about it for our sake! We'll get it in eventually :)

**Comment by andrewthauer** [comment]
We removed the explore plugin because it was providing little value in our instance. I like the idea of explicitly making this more task oriented. As pointed out previously there will be multiple personas and this use case is probably served mostly by someone exploring rather then going directly. I'm guessing the actual domain pages will still exist within the catalog routing?

**Comment by freben** [comment]
> I'm guessing the actual domain pages will still exist within the catalog routing?

I'd argue they belong to the catalog, yeah.

**Comment by Fox32** [comment]
Yes, the explorer links to the entity pages of the domains. It's just a way to see the existing domains because they aren't display in the catalog table (right now).

**Comment by Fox32** [comment]
Made this a draft for now because I'm working on it. 

Good news: ported the explore plugin to the new routing and it worked!!!! 🎉

**Comment by Fox32** [comment]
@Rugvip For the tools tab in the explore plugin: I wonder whether the new extension system brings something where I get the entry point route of every plugin (in case a plugin has something like that) or list all available plugins? 

I could imaging to display all plugins that have a root route, e.g. the catalog, lighthouse, ... as card here?

Any pointers? If it doesn't exist yet, I will just create a simple API that integrators can override with static data.

File: plugins/explore/src/components/ToolCard/ToolCard.tsx
Comment (Fox32) [inline plugins/explore/src/components/ToolCard/ToolCard.tsx:32]: I really like the tool card styling (tags at the bottom, the ability to add a picture alter) but I prefer to go with the build in style from storybook for both cards. Maybe we improve the `ItemCard`?

File: plugins/explore/src/components/ToolExplorerContent/ToolExplorerContent.tsx
Comment (Fox32) [inline plugins/explore/src/components/ToolExplorerContent/ToolExplorerContent.tsx:20]: If there is no way to access the available plugins, I would provide a static API that can be used to inject a custom list

**Comment by Fox32** [comment]
Updated this a bit, not 100% finished yet but I would look forward to some feedback

image


image

**Comment by Fox32** [comment]
This is how it looks without having the padding twice 😉 
image

**Comment by Fox32** [comment]
I added an API to configure the available tools statically.

Now the last thing that is missing is an overhaul of the item card, but I still think that it is ready for review.
The item card looks different for tools and domains. Personally I prefer the tools card, with some minor changes (e.g. fallback if image is missing), but the one from domains is in storybook. My suggestion is to keep it as-is a rework the item card in storybook. Or we decide to not use the one from storybook and reuse the tools card for the domains.

File: packages/app/src/plugins.ts
Comment (OrkoHunter) [inline packages/app/src/plugins.ts:21]: Nit - Looks a bit different than the rest. :)

File: packages/app/src/App.tsx
Comment (OrkoHunter) [inline packages/app/src/App.tsx:41]: I like this grouping of imports and I try to do it too. Do you have an extension which can do this in your editor, or was this a manual attempt?

File: packages/core/src/layout/ItemCard/ItemCard.tsx
Comment (OrkoHunter) [inline packages/core/src/layout/ItemCard/ItemCard.tsx:35]: Do these numbers have meanings? Can they be variables? The calc seems unnecessary as it is - if it can just be `135px` and a comment can explain why.

**Review by OrkoHunter — APPROVED** [review]
Approving for TechDocs changes. 👌

File: packages/app/src/plugins.ts
Comment (Fox32) [reply packages/app/src/plugins.ts:21]: Followed this: 

image

https://backstage.io/docs/plugins/composability#naming-patterns

File: packages/app/src/App.tsx
Comment (Fox32) [reply packages/app/src/App.tsx:41]: I don't think it's an extension, I run "Organize Imports" followed by "Format Document" (Prettier) in VSCode

File: packages/app/src/plugins.ts
Comment (OrkoHunter) [reply packages/app/src/plugins.ts:21]: Aha! Thanks, I didn't realize the naming pattern change. 🙏

File: packages/core/src/layout/ItemCard/ItemCard.tsx
Comment (Fox32) [reply packages/core/src/layout/ItemCard/ItemCard.tsx:35]: Yes, the are related to the paddings in `content` and `header`, but not sure how I could pull this into variables as `theme.spacing` is used there and the spacing is not a pixel value.

File: packages/app/src/App.tsx
Comment (freben) [reply packages/app/src/App.tsx:41]: `shift + option + O` gang unite, i do this compulsively

Comment (freben) [reply packages/app/src/App.tsx:41]: (but it does lead to messy diffs ;))

File: packages/app/src/plugins.ts
Comment (freben) [reply packages/app/src/plugins.ts:21]: image

File: packages/app/src/App.tsx
Comment (OrkoHunter) [reply packages/app/src/App.tsx:41]: First diff messy, much cleaner afterwords. I'm in.

File: packages/core/src/layout/ItemCard/ItemCard.tsx
Comment (freben) [inline packages/core/src/layout/ItemCard/ItemCard.tsx:78]: `classnames` or `clsx` (see elsewhere in the repo for usages) are great for this kind of thing

File: plugins/explore/README.md
Comment (freben) [inline plugins/explore/README.md:11]:
```suggestion
export { plugin as Explore } from '@backstage/plugin-explore';
```

File: plugins/explore/src/components/DomainCard/DomainCardGrid.tsx
Comment (freben) [inline plugins/explore/src/components/DomainCard/DomainCardGrid.tsx:22]: At this point i started thinking about what it MEANS to import a non-versioned kind :) I think later on, we'll have to start actually settling this and deciding who has to normalize the schemas - the writer side (so that the catalog only ever STORES one version of schema), or the reader side

Comment (freben) [reply plugins/explore/src/components/DomainCard/DomainCardGrid.tsx:22]: (never mind this now ... but at some point we need to talk about it)

File: plugins/explore/src/components/ExplorePage/ExploreTabs.tsx
Comment (freben) [reply plugins/explore/src/components/ExplorePage/ExploreTabs.tsx:22]: The `EntityPageLayout` found in the catalog uses the new pattern. We'll probably want to make a core variant of that that everybody can benefit from.

File: plugins/explore/src/components/ToolCard/ToolCard.tsx
Comment (freben) [reply plugins/explore/src/components/ToolCard/ToolCard.tsx:32]: Yeah we should unify this later and make them great.

File: plugins/explore/src/components/ToolExplorerContent/ToolExplorerContent.tsx
Comment (freben) [inline plugins/explore/src/components/ToolExplorerContent/ToolExplorerContent.tsx:26]: oops, should be a module import

**Review by freben — COMMENTED** [review]
Cool stuff!

File: plugins/explore/src/components/ToolExplorerContent/ToolExplorerContent.tsx
Comment (Fox32) [reply plugins/explore/src/components/ToolExplorerContent/ToolExplorerContent.tsx:26]: 😱  Hmm, why doesn't any tooling detected this... 😆

Comment (Fox32) [reply plugins/explore/src/components/ToolExplorerContent/ToolExplorerContent.tsx:26]: VSCode loves to do it when moving stuff around...

Comment (freben) [reply plugins/explore/src/components/ToolExplorerContent/ToolExplorerContent.tsx:26]: Yeah I know. I thought we had a lint check for this, actually. I must be misremembering

Comment (freben) [reply plugins/explore/src/components/ToolExplorerContent/ToolExplorerContent.tsx:26]: No we do!

```ts
    // Avoid cross-package imports
    'no-restricted-imports': [2, { patterns: ['**/../../**/*/src/**'] }],
```

wonder why that didn't fire ...

File: plugins/explore/README.md
Comment (Fox32) [reply plugins/explore/README.md:11]: Even more wrong, should be `explorePlugin` according to Mando 😆

File: packages/core/src/layout/ItemCard/ItemCard.tsx
Comment (Fox32) [reply packages/core/src/layout/ItemCard/ItemCard.tsx:78]: Choose `clsx` as it comes with typings out of the box. `@backstage/code` as both dependencies, maybe we should settle on one?

File: packages/core/src/layout/ItemCard/ItemCard.tsx
Comment (Fox32) [reply packages/core/src/layout/ItemCard/ItemCard.tsx:89]: This looks kind of strange when you hover because the button text is underlined. When you don't hover it looks the same as a raw `<Link />`, maybe it has some more padding.

File: plugins/explore/src/components/ToolExplorerContent/ToolExplorerContent.tsx
Comment (Fox32) [reply plugins/explore/src/components/ToolExplorerContent/ToolExplorerContent.tsx]: Maybe because the `/**` at then end?

Comment (freben) [reply plugins/explore/src/components/ToolExplorerContent/ToolExplorerContent.tsx]: yeah that sounds sensible

File: plugins/explore/src/components/ExplorePage/ExploreTabs.tsx
Comment (Fox32) [reply plugins/explore/src/components/ExplorePage/ExploreTabs.tsx:22]: I will extract it into `@backstage/core`

Comment (Fox32) [reply plugins/explore/src/components/ExplorePage/ExploreTabs.tsx:22]: Move it to @backstage/core and it works quite well. But I'm unable to display it in storybook. What kind of wrapper do I need around it? It can't discover the `Component.Route` as children.

Comment (Fox32) [reply plugins/explore/src/components/ExplorePage/ExploreTabs.tsx]: Found it, looks like `child.type !== Route` isn't always a safe operation, especially in storybook, but I found a replacement.

File: plugins/explore/src/components/ToolCard/ToolCard.test.js
Comment (andrewthauer) [inline plugins/explore/src/components/ToolCard/ToolCard.test.js:17]: This file seems to be us `.js` vs `.ts` for some reason

Comment (Fox32) [reply plugins/explore/src/components/ToolCard/ToolCard.test.js:17]: Yeah, this seems to be a very old file 😆 Not sure if we want to fix that, just changing the extension is not enough. I will have a look how much work it is.

Comment (Fox32) [reply plugins/explore/src/components/ToolCard/ToolCard.test.js]: Works!

**Comment by Fox32** [comment]
This monster should be ready now, it got quite big...

**Comment by Fox32** [comment]
@freben if the PR is to big, I could consider landing the `ItemCard` and `TabbedLayout` changes in separate PRs, what do you think?

**Comment by freben** [comment]
Let me have a look, let's see

**Review by freben — APPROVED** [review]
This was super solid work, good going!

File: plugins/explore-react/src/tools/api.ts
Comment (Fox32) [reply plugins/explore-react/src/tools/api.ts:35]: Would be quite neat if it was part of the config. You could link in a yaml file into the config and display it. But I think we have to view this from the use case of the data: If the data is potentially available in a Backend the API would be a good choice. But I would doubt that somebody has this in a backend 😆 On the other hand my goal for that page would be do display the root routes for all backstage plugins. The API is just a temporary solution, there is a lot more room to improve. I only added it because I wanted to hide the static data.

**Comment by regicsolutions** [comment]
@Fox32 is there an example of how i can pull in the tools samples? domains is working as expected but not the tools tab under explore/tools

**Comment by andrewthauer** [comment]
@regicsolutions - Here is the example data structure: https://github.com/backstage/backstage/blob/master/plugins/explore/src/util/examples.ts.

Then add the following to the `packages/app/src/apis.ts` file to use it:

```ts
import { exploreToolsConfigRef } from '@backstage/plugin-explore-react';

export const apis = [
  // ...
  createApiFactory({
    api: exploreToolsConfigRef,
    deps: {},
    factory: () => ({
      async getTools() {
        return exploreTools; // the tools array data goes here
      },
    }),
  }),
```

Probably should be documented though :)

**Comment by regicsolutions** [comment]
> @regicsolutions - Here is the example data structure: https://github.com/backstage/backstage/blob/master/plugins/explore/src/util/examples.ts.
> 
> 
> 
> Then add the following to the `packages/app/src/apis.ts` file to use it:
> 
> 
> 
> ```ts
> 
> import { exploreToolsConfigRef } from '@backstage/plugin-explore-react';
> 
> 
> 
> export const apis = [
> 
>   // ...
> 
>   createApiFactory({
> 
>     api: exploreToolsConfigRef,
> 
>     deps: {},
> 
>     factory: () => ({
> 
>       async getTools() {
> 
>         return exploreTools; // the tools array data goes here
> 
>       },
> 
>     }),
> 
>   }),
> 
> ```
> 
> 
> 
> Probably should be documented though :)

Thanks!

**Comment by regicsolutions** [comment]
@Fox32 Great work here! what are your thoughts on adding actions to each tool to handle links to different environments? Found it to be a useful feature that the tech stacks plugin offered - https://github.com/easylo/backstage/blob/issue-734/plugins/tech-stack/src/config/ToolsListConfig.ts

**Comment by Fox32** [comment]
I didn't made any functional changes to the tools list, I just added the domains here.

> One other suggestion is to force a max width and height for images

One follow up task would be to define a better ItemCard with image support.

**Comment by ArchanaSwamy** [comment]
> @regicsolutions - Here is the example data structure: https://github.com/backstage/backstage/blob/master/plugins/explore/src/util/examples.ts.
> 
> Then add the following to the `packages/app/src/apis.ts` file to use it:
> 
> ```ts
> import { exploreToolsConfigRef } from '@backstage/plugin-explore-react';
> 
> export const apis = [
>   // ...
>   createApiFactory({
>     api: exploreToolsConfigRef,
>     deps: {},
>     factory: () => ({
>       async getTools() {
>         return exploreTools; // the tools array data goes here
>       },
>     }),
>   }),
> ```
> 
> Probably should be documented though :)

This needs a correction and I am having an issue as well. It should be 
 return exampleTools;

Also what should be the expected `import` statement for this?

**Comment by ArchanaSwamy** [comment]
@regicsolutions Wondering if you  got the  tools section working.  I am having the same problem despite adding the above details in the `api.ts`

**Comment by andrewthauer** [comment]
Here is the full example. Hopefully this helps:

```ts
// in packages/app/src/plugins/explore.ts (new)
export const exploreTools = [
  {
    title: 'Backstage',
    description: 'Open source platform for building developer portals.',
    url: 'https://backstage.io/',
    image: 'https://example/some-image.png',
    tags: ['dev', 'portal'],
  },
};

// in packages/app/src/apis.ts
import { exploreTools } from './plugins/explore';

export const apis = [
  // ...
  createApiFactory({
    api: exploreToolsConfigRef,
    deps: {},
    factory: () => ({
      async getTools() {
        return exploreTools;
      },
    }),
  }),
  // ...
];
```

**Comment by andrewthauer** [comment]
Also, another thing. I think this only works if you convert the `packages/app/src/App.tsx` routes over to use `FlatRoutes`. See https://github.com/backstage/backstage/blob/master/packages/app/src/App.tsx#L69.

**Comment by ArchanaSwamy** [comment]
Thank you @andrewthauer . Fixing the `FlatRoutes` worked.

**Comment by ArchanaSwamy** [comment]
Any idea what is the  purpose of `newsTag` in the exploreToolsConfigRef?

**Comment by afadil** [comment]
Here is an example to fetch tools list from the catalog:

// in packages/app/src/plugins/explore.ts (new)
export const exploreTools = [
  {
    title: 'Backstage',
    description: 'Open source platform for building developer portals.',
    url: 'https://backstage.io/',
    image: 'https://example/some-image.png',
    tags: ['dev', 'portal'],
  },
};

// in packages/app/src/apis.ts
import { exploreTools } from './plugins/explore';

export const apis = [
  // ...
   createApiFactory({
    api: exploreToolsConfigRef,
    deps: { catalogApi: catalogApiRef },
    factory: ({ catalogApi }) => ({
      async getTools() { 
        return catalogApi.getEntities({
          filter: { kind: 'component', 'spec.type': 'tool' },
        }).then((entities) => {
          return entities.items.map((entity) => {
            return {
              title: entity.metadata.title,
              description: entity.metadata.description,
              url: entity.metadata.name,
              image: entity?.spec?.profile?.image,
              icon: entity?.spec?.profile?.icon,
              tags: entity.metadata.tags,
            };
          });
        });
      }
    }),
  }),
  // ...
];

**Pull request merged**
